blog/content/post/2023-09-27-aem_phase2.md: add blog post (#543)
3mdeb/news-and-ideas: blog/content/post/2023-09-27-aem_phase2.md

  - Hypervisor
  - OS Dev
  - Security

---

# Introduction

If you haven't read the previous [TrenchBoot AEM blog
post](https://blog.3mdeb.com/2023/2023-01-31-trenchboot-aem-for-qubesos/) or
our [project plan](https://docs.dasharo.com/projects/trenchboot-aem-v2/) yet,
it may be a good lecture to catch up on what the TrenchBoot AEM project is
about. This post will guide you through the most recent progress and show a
demo of TrenchBoot AEM in action, plus installation steps.

## What's new in TrenchBoot AEM?

The main focus of this phase was to get the TPM 2.0 working with Qubes OS AEM
scripts on the DRTM boot flow in TrenchBoot (GRUB and Xen). Let's go through
each of the tasks quickly.

### Implement parallel CPU cores bring-up for DRTM launch

It is not strictly related to TPM 2.0 support, but after phase 1, it occurred
to be necessary to clean up the code for CPU bring-up in Xen. As the phase 1
was just a Proof of Concept, some inefficiencies in the code happened to land.
The most significant change in this [pull
request](https://github.com/TrenchBoot/xen/pull/1/files) is the parallel
bring-up of the CPUs, in opposite to the previous implementation, which caused

> the CPUs to wait in a queue. It had to be fixed because if an interrupt
> occurred during the wait time, it could pose a severe threat to security.

Reviewer: ```suggestion
the CPUs to wait in a queue. It had to be fixed because PoC implementation
hardcoded some of CPU features as enabled, which could potentially break booting
without DRTM.
```

### Support for TPM 2.0 module in Xen

With the introduction of TPM 2.0 support, Xen also required changes that would
perform the dom0 kernel and initrd measurements properly. As a quick reminder,
TPM 1.2 and TPM 2.0 communication protocols to extend the PCRs with software
measurements differ slightly. Also, TPM 2.0 uses stronger hash algorithms than
TPM 1.2, like SHA256, which had to be added to Xen. Changes are outlined in
this [pull request](https://github.com/TrenchBoot/xen/pull/3/files).

### Support for TPM 2.0 event log in Xen

Similarly to the previous task, besides the communication protocol difference,
the event log, where the information about measurements is stored, differs
between TPM 1.2 and TPM 2.0. As a next step after measuring the dom0 kernel
and initrd to TPM 2.0 PCRs, we had to record the hashes in the event log in
TPM 2.0 compliant format. The event log is a very important chunk of
information, which lets one confirm what is measured and if the TPM PCRs

> actually match the values that have been extended (basically an attestation).

Reviewer: False. You can't trust event log unless you know that final PCR values are correct. Without it, anything with event log matching final values would be considered as trusted, but at best it only means that the code producing the log is proper.

```suggestion
actually match the values that have been extended.
```

---
title: TrenchBoot Anti Evil Maid - Phase 2
abstract: 'TrenchBoot Anti Evil Maid project for Qubes OS is progressing. With
           the addition of TPM 2.0 support, Anti Evil Maid gains much higher
           adoption and possibilities than ever before.'
cover: /covers/trenchboot-logo.png
author: michal.zygowski
layout: post
published: true

> date: 2023-09-27

Author: Date is always updated when @artur-rs publishes a blog post by merging develop to master

archives: "2023"

tags:
  - coreboot
  - QubesOS
  - TrenchBoot
  - GRUB2
  - Xen
  - Hypervisor
categories:
  - Firmware
  - Bootloader
  - Hypervisor
  - OS Dev
  - Security

---

# Introduction

If you haven't read the previous [TrenchBoot AEM blog
post](https://blog.3mdeb.com/2023/2023-01-31-trenchboot-aem-for-qubesos/) or
our [project plan](https://docs.dasharo.com/projects/trenchboot-aem-v2/) yet,
it may be a good lecture to catch up on what the TrenchBoot AEM project is
about. This post will guide you through the most recent progress and show a

> demo of TrenchBoot AEM in action, plus installation steps.

Author: https://github.com/3mdeb/news-and-ideas/pull/543/commits/23be3297f4ca42f04cfc9231a9242525abf008d2#diff-c2e0e25a477d5cc5bf3075df551332eee15aaba457379ef35e8547639f285757R39

second time to provide any of the disk passwords for confirmation.

The system now is ready to perform AEM boot, so reboot your machine!
Now you should be presented with a SRK password prompt like this:

![AEM SRK password prompt](/img/qubes_aem_v0.2_13.png)

And later with the standard disk password prompt:

![AEM disk password prompt](/img/qubes_aem_v0.2_14.png)

After that AEM will print the messages saying that the secrets have been
sealed to the LUKS PCR 13 DRTM PCRs 17 and 18:

![AEM secrets sealing](/img/qubes_aem_v0.2_15.png)

You should see the above screen only once (unless you change something with
GRUB, Xen or dom0 kernel/initrd). Now that the AEM knows what the final PCR
values should be, after next reboot AEM will successfully unseal the secrets
it just sealed before a moment, and present a TOTP code. But before that, save
the current PCR values by invoking the following command in dom0 terminal:

```bash
sudo tpm2_pcrread sha256
```

![TPM PCRs](/img/qubes_aem_v0.2_16.png)

Note the PCR values with `sudo tpm2_pcrread sha256 > drtm_pcrs.txt` for
example and reboot the machine. If everything went correctly, the PCRs should

> not change after reboot (you can comapre by reading the PCRs again with

Author: This sentence has been removed: https://github.com/3mdeb/news-and-ideas/pull/543/commits/23be3297f4ca42f04cfc9231a9242525abf008d2#diff-c2e0e25a477d5cc5bf3075df551332eee15aaba457379ef35e8547639f285757L330

`tpm2_pcrread` to a file and `diff` command when booted) and you should be
presented a TOTP code during LUKS key file password prompt, like this:

![AEM TOTP code](/img/qubes_aem_v0.2_17.png)

Check the TOTP code in your authenticator app to ensure it is correct and then
type the password provided in the `anti-evil-maid-install` step. If there was
any error, it should be displayed in the top left corner of the screen and you
would be prompted for your regular disk password, because the LUKS key would
not be unsealed by AEM.

### Troubleshooting

Sometimes the TOTP code does not immediately appear after second reboot,
because the AEM secret could not be unsealed. It may be caused by the TPM

> safety timer. If your machine was not safely shut down (e.g. due to a power loss) and it had been running for less than approximately 70 minutes

Reviewer: Wrap

Author: https://github.com/3mdeb/news-and-ideas/pull/543/commits/23be3297f4ca42f04cfc9231a9242525abf008d2#diff-c2e0e25a477d5cc5bf3075df551332eee15aaba457379ef35e8547639f285757R342

(2^22 ms) since, you may experience such issue. Simply try to reboot the machine at
a later time and TOTP should appear.

You can check the AEM services status using the following command in dom0:

```bash
journalctl -u anti-evil-maid-unseal -u anti-evil-maid-seal
```

## Summary

It has been quite a long journey implementing the TPM 2.0 support, yet much
more awaits us. We bet you are eager to hear more about TrenchBoot AEM and
can't wait to see the AMD DRTM or UEFI boot mode support coming to Qubes OS.
It will be the first time AEM is launched on an AMD platform, which certainly
stirs up some hype. If you like what we do, please consider supporting us:

* if you can hack a bit, your testing results will be invaluable,
* spread the word about us and the project on social media,
* any other way you see fit.

If you think we can help in improving the security of your firmware or you are
looking for someone who can boost your product by leveraging advanced features
of used hardware platform, feel free to
[book a call with us](https://calendly.com/3mdeb/consulting-remote-meeting) or
drop us email to `contact<at>3mdeb<dot>com`. If you are interested in similar
content, feel free to

> [sign up to our newsletter](https://newsletter.3mdeb.com/subscription/PW6XnCeK6)

Author: https://github.com/3mdeb/news-and-ideas/pull/543/commits/23be3297f4ca42f04cfc9231a9242525abf008d2#diff-c2e0e25a477d5cc5bf3075df551332eee15aaba457379ef35e8547639f285757R371

### Support for TPM 2.0 module in Xen

With the introduction of TPM 2.0 support, Xen also required changes that would

> perform the Dom0 kernel and initrd measurements properly. As a quick reminder,

Author: Because one of the requirements has Dom0 with capital letter:

`Integrate TPM 2.0 software stack into Qubes OS Dom0`

Had to change it for the sake of consistence

> perform the Dom0 kernel and initrd measurements properly. As a quick reminder,

Reviewer: I hardly find requirements defined in MoU as source of truth on that matter.

> perform the Dom0 kernel and initrd measurements properly. As a quick reminder,

Author: 7726bb3ead3040889f32aa4274105c76061f8632

$ gpg --verify sha512sums.sig sha512sums
gpg: Signature made Thu 19 Oct 2023 01:55:19 PM CEST
gpg:                using RSA key 3405D1E4509CD18A3EA762245D289020C07114F3
gpg: Good signature from "Qubes OS TrenchBoot AEM open-source software release 0.x signing key" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 3405 D1E4 509C D18A 3EA7  6224 5D28 9020 C071 14F3

$ sha512sum -c sha512sums
anti-evil-maid-4.1.1-1.fc37.x86_64.rpm: OK
grub2-common-2.06-2.fc37.noarch.rpm: OK
grub2-pc-2.06-2.fc37.x86_64.rpm: OK
grub2-pc-modules-2.06-2.fc37.noarch.rpm: OK
grub2-tools-2.06-2.fc37.x86_64.rpm: OK
grub2-tools-extra-2.06-2.fc37.x86_64.rpm: OK
grub2-tools-minimal-2.06-2.fc37.x86_64.rpm: OK
python3-xen-4.17.2-4.fc37.x86_64.rpm: OK
xen-4.17.2-4.fc37.x86_64.rpm: OK
xen-hypervisor-4.17.2-4.fc37.x86_64.rpm: OK
xen-libs-4.17.2-4.fc37.x86_64.rpm: OK
xen-licenses-4.17.2-4.fc37.x86_64.rpm: OK
xen-runtime-4.17.2-4.fc37.x86_64.rpm: OK
```

Our starting point is the fresh installation of [Qubes OS
R4.2.0-rc3](https://ftp.qubes-os.org/iso/Qubes-R4.2.0-rc3-x86_64.iso) with the

> AEM v0.2 packages and Intel SINIT ACMs placed in Dom0 home directory.

Author: If I prefix it with TrenchBoot, would that suffice?

> AEM v0.2 packages and Intel SINIT ACMs placed in Dom0 home directory.

Reviewer: Yeah, it should be enough.

> AEM v0.2 packages and Intel SINIT ACMs placed in Dom0 home directory.

Author: 6ba7e4800b92400f6b56c9ae8fdf49a49f0b00e6

To download SINIT ACMs from Intel portal, simply invoke:

```bash
wget -O 630744_003.zip https://cdrdv2.intel.com/v1/dl/getContent/630744
```

> Assuming that the RPMs and ZIP is in your Dom0 home directory already, open

Reviewer: This was undone by https://github.com/3mdeb/news-and-ideas/commit/7726bb3ead3040889f32aa4274105c76061f8632

> Assuming that the RPMs and ZIP is in your Dom0 home directory already, open

Author: my bad during rebase

the Dom0 terminal. Unzip the ACMs package and copy the right ACM to the boot
partition. In our case it is:

```bash
sudo cp 630744_003/SKL_KBL_AML_SINIT_20211019_PRODUCTION_REL_NT_O1_1.10.0.bin /boot
```

Next, install required dependencies for AEM (we do it manually because we
install local RPM packages later):

```bash
sudo qubes-dom0-update --enablerepo=qubes-dom0-current-testing \
    oathtool \
    openssl \
    qrencode \
    tpm-extra \
    trousers-changer \
    tpm-tools
```

![Installing dependencies 1](/img/qubes_aem_v0.2_1.png)

> ![Installing dependencies 2](/img/qubes_aem_v0.2_2.png)

Author: 7feeed2b99a4266e4e8ed61d8b5c7e2e2e868fbf

Qubes OS R4.2.0-rc3):

```bash
sudo dnf reinstall \
	./grub2-common-2.06-2.fc37.noarch.rpm \
	./grub2-pc-2.06-2.fc37.x86_64.rpm \
	./grub2-pc-modules-2.06-2.fc37.noarch.rpm \
	./grub2-tools-2.06-2.fc37.x86_64.rpm \
	./grub2-tools-minimal-2.06-2.fc37.x86_64.rpm
```

![Reinstalling GRUB packages 1](/img/qubes_aem_v0.2_5.png)
![Reinstalling GRUB packages 2](/img/qubes_aem_v0.2_6.png)

Before installing AEM package, we also need to perform GRUB installation to
the disk where Qubes OS is stored, in our case it is `/dev/sda` (yours may be
different so watch out!). This will ensure the support for DRTM is installed
in GRUB:

```bash
sudo grub2-install /dev/sda
```

At last we can install the AEM package itself:

```bash
sudo dnf install ./anti-evil-maid-4.1.1-1.fc37.x86_64.rpm
```
> When installing AEM package it may seem that it hung at `Running scriptlet`
> step. Please be patient and let the system finish installation.

![Installing GRUB to disk and AEM package](/img/qubes_aem_v0.2_7.png)
![Installing GRUB to disk and AEM package](/img/qubes_aem_v0.2_8.png)

All packages are in place. Before we can proceed with provisioning AEM, we

> have to clear the TPM in the BIOS. Otherwise the next step may return the

Author: > IIUC TPM must be reset only if the ownership was previously taken, which in most cases shouldn't be.

Well, you would probably be surprised. I have almost never seen a situation where I do not have to clear TPM when provisioning AEM.

> have to clear the TPM in the BIOS. Otherwise the next step may return the

Author: Reworked the order per your suggestion 50ce7d892753c88c9a6998c16a3a73a4cdf9b129

following error:

![AEM TPM setup error](/img/qubes_aem_v0.2_9.png)

Some platforms may require disabling Intel Trusted Execution Technology (TXT)
in order to clear TPM. So after you clear the TPM, remember to enable Intel
TXT back, otherwise AEM will not work. Once TPM is cleared, perform the TPM
setup:

```bash
sudo anti-evil-maid-tpm-setup
```

![AEM TPM setup 1](/img/qubes_aem_v0.2_10.png)

You will be prompted to set the SRK password, it is a password to access TPM's
nonvolatile storage where the AEM secrets will be sealed.

I have also shown the TPM properties to prove that we are using TPM 2.0. You
can omit this. Then let's perform the AEM installation. In our example we are
installing the AEM with MFA to the internal disk where Qubes OS is installed.
This is generally not recommended, because using MFA and internal disk does
not provide any extra security over not using MFA. But it is just for
demonstration purposes. So in your case feel free to install AEM even on an

> external storage. In our case the boot partition is `/dev/sda1` (your may be

Reviewer: Why have you marked this as resolved? Are you trying to hide this?

> external storage. In our case the boot partition is `/dev/sda1` (your may be

Author: My bad during rebase. Also committing suggestions via GUI automatically resolves discussion, sorry

> external storage. In our case the boot partition is `/dev/sda1` (your may be

Author: https://github.com/3mdeb/news-and-ideas/pull/543/commits/23be3297f4ca42f04cfc9231a9242525abf008d2#diff-c2e0e25a477d5cc5bf3075df551332eee15aaba457379ef35e8547639f285757R292

authentication app of your choice. it will show you a TOTP code later at boot
process. This step will also ask you to provide a password for the secret and
second time to provide any of the disk passwords for confirmation.

The system now is ready to perform AEM boot, so reboot your machine!
Now you should be presented with a SRK password prompt like this:

![AEM SRK password prompt](/img/qubes_aem_v0.2_13.png)

And later with the standard disk password prompt:

![AEM disk password prompt](/img/qubes_aem_v0.2_14.png)

After that AEM will print the messages saying, that the secrets have been
sealed to the LUKS PCR 13 DRTM PCRs 17 and 18:

![AEM secrets sealing](/img/qubes_aem_v0.2_15.png)

You should see the above screen only once (unless you change something with
GRUB, Xen or Dom0 kernel/initrd). Now that the AEM knows what the final PCR
values should be, after next reboot AEM will successfully unseal the secrets
it just sealed before a moment, and present a TOTP code. But before that, save
the current PCR values by invoking the following command in Dom0 terminal:

```bash
sudo tpm2_pcrread sha256
```

![TPM PCRs](/img/qubes_aem_v0.2_16.png)

> Note the PCR values somewhere and reboot the machine. If everything went

Author: https://github.com/3mdeb/news-and-ideas/pull/543/commits/50ce7d892753c88c9a6998c16a3a73a4cdf9b129#diff-c2e0e25a477d5cc5bf3075df551332eee15aaba457379ef35e8547639f285757R328

> Note the PCR values somewhere and reboot the machine. If everything went

Reviewer: This doesn't change the meaning, now it even more suggests that saving those values is required. AEM will work without it. This may serve as debugging aid if something doesn't work, but right now it only makes the instructions more complicated. Please either remove it, move this to `Troubleshooting` or explicitly say that this is optional.

> Note the PCR values somewhere and reboot the machine. If everything went

Author: Removed: https://github.com/3mdeb/news-and-ideas/pull/543/commits/23be3297f4ca42f04cfc9231a9242525abf008d2#diff-c2e0e25a477d5cc5bf3075df551332eee15aaba457379ef35e8547639f285757L323

> correctly, the PCRs should not change after reboot (you can comapre by reading

Reviewer: ```suggestion
correctly, the PCRs should not change after reboot (you can compare by reading
```

Author: This sentence has been removed: https://github.com/3mdeb/news-and-ideas/pull/543/commits/23be3297f4ca42f04cfc9231a9242525abf008d2#diff-c2e0e25a477d5cc5bf3075df551332eee15aaba457379ef35e8547639f285757L330

the PCRs again with `tpm2_pcrread` when booted) and you should be presented a
TOTP code during LUKS key file password prompt, like this:

![AEM TOTP code](/img/qubes_aem_v0.2_17.png)

Check the TOTP code in your authenticator app to ensure it is correct and then

> type the password provided in the `anti-evil-maid-install` step. if there was

Reviewer: ```suggestion
type the password provided in the `anti-evil-maid-install` step. If there was
```
